### PR TITLE
fix #149 add mutex lock/unlock

### LINKED ISSUE #149: Panic after calling Send (after x hours of functioning properly)
I'm not sure if this is a bug in the SDK (it looks similar to this issue: https://github.com/Azure/azure-service-bus-go/issues/141), or a implementation issue at my side that I'm not aware of, but after hours of functioning properly and publishing messages to the Azure Service Bus, my application panic-ed with the following trace:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xb2110d]

goroutine 1072417 [running]:
github.com/Azure/azure-service-bus-go.(*Sender).Send(0xc00019f260, 0x1620840, 0xc0002b43c0, 0xc0008ab3b0, 0x0, 0x0, 0x0, 0x0, 0x0)
/go/pkg/mod/github.com/!azure/azure-service-bus-go@v0.9.1/sender.go:151 +0x32d

github.com/Azure/azure-service-bus-go.(*Topic).Send(0xc0003978e0, 0x1620840, 0xc0002b43c0, 0xc0008ab3b0, 0x0, 0x0, 0x0, 0x0, 0x0)
/go/pkg/mod/github.com/!azure/azure-service-bus-go@v0.9.1/topic.go:105 +0x1c1

github.com/brocaar/lora-app-server/internal/integration/azureservicebus.(*Integration).publish(0xc0002b4300, 0x1620900, 0xc000a9a900, 0x135d2cd, 0x2, 0x18, 0x270375d39373433, 0x131a280, 0xc0007c59a0, 0x0, ...)
/lora-app-server/internal/integration/azureservicebus/azure_service_bus.go:193 +0x6b9

github.com/brocaar/lora-app-server/internal/integration/azureservicebus.(*Integration).SendDataUp(0xc0002b4300, 0x1620900, 0xc000a9a900, 0x18, 0xc00098f6e0, 0x13, 0xc000534aa0, 0x10, 0x270375d39373433, 0xc000855a70, ...)
/lora-app-server/internal/integration/azureservicebus/azure_service_bus.go:78 +0xd7
```

As you can see in the trace, this is where the SDK gets called:
https://github.com/brocaar/lora-app-server/blob/v3.4.0/internal/integration/azureservicebus/azure_service_bus.go#L193

Before the panic, `Send` did publish a message to the Service Bus without error (e.g. there was no indication that there was a connection issue). This was in the logfiles before the panic:

```
integration/azureservicebus: event published" ctx_id=xxxxxxxx dev_eui=xxxxxxx event=up
```

(https://github.com/brocaar/lora-app-server/blob/v3.4.0/internal/integration/azureservicebus/azure_service_bus.go#L199)

### PR DESCRIPTION
**Pull request by johejo**
fix #149

### CODE REVIEW

**Comment by johejo** [comment]
Hi maintainers, Can you review this PR?
Panic is frequent in our production environment ... :cry: .

File: sender.go [hunk 1]
```diff
@@ -147,6 +147,9 @@ func (s *Sender) Send(ctx context.Context, msg *Message, opts ...SendOption) err
 	ctx, span := s.startProducerSpanFromContext(ctx, "sb.Sender.Send")
 	defer span.End()
 
+	s.clientMu.Lock()
```
Comment (jhendrixMSFT) [inline sender.go]: From the discussion in the issue I thought there was a risk of deadlock with this fix.  @catalinaperalta and I had tried acquiring/releasing the lock if `msg.SessionID` is nil (see [here](https://github.com/Azure/azure-service-bus-go/issues/149#issuecomment-557659831)) however in some of our testing it appeared to introduce a bit of a perf regression.  If the perf issue is real I would expect this change to exhibit the same behavior.  Have you observed this?
At any rate I would prefer scoping the lock acquisition/release to the below `if` block unless it's incomplete.

Comment (johejo) [reply sender.go]: Since Go1.14 , the performance of defer has improved significantly.
https://golang.org/doc/go1.14#runtime

Comment (jhendrixMSFT) [reply sender.go]: I don't think the perf had anything to do with defer although we didn't measure that specifically.  Regardless I think it's best to hold the read lock for the shortest duration required.  With the fix as-is, the read lock is held until Send() exits which seems excessive (unless we know it's a requirement for the fix to properly work).

Comment (johejo) [reply sender.go]: Thanks, I got it.
I'll try to fix.

P.S.
I found there is mutex performance improvement in Go1.14.
https://golang.org/doc/go1.14#library

**Comment by jhendrixMSFT** [comment]
@johejo thanks for making the changes.  Once I get confirmation from you that this resolves your issue with no unwanted side-effects I will merge and tag this PR.

**Comment by johejo** [comment]
Hi, I deployed this branch with go.mod replace tag in my environment since last week, there is no panic.
Also the performance seems to be no changed.

File: sender.go [hunk 2]
```diff
@@ -148,8 +148,10 @@ func (s *Sender) Send(ctx context.Context, msg *Message, opts ...SendOption) err
 	defer span.End()
 
 	if msg.SessionID == nil {
+		s.clientMu.Lock()
```
Comment (jhendrixMSFT) [inline sender.go]: This should be the read lock, i.e. `s.clientMu.RLock()` and use `RUnlock()` below.

**Comment by jhendrixMSFT** [comment]
@johejo thanks for testing.  Unfortunately the change is taking the write lock, not the read lock.  Would you be able to give it one more try taking the read lock?

File: sender.go (hunk 2 above)
Comment (johejo) [reply sender.go]: Oh, my mistake.

**Comment by princjef** [comment]
is this ready to be merged? the lock mechanism looks to have been corrected to `RLock`/`RUnlock`

**Comment by jhendrixMSFT** [comment]
I was waiting for confirmation from @johejo that switch to the read-lock resolved the issue.  If it did then I will merge and tag.

**Comment by johejo** [comment]
Unfortunately the nil dereference was not resolved.
Crashes sometimes occur.
I've been looking up for another solution, giving up from this problem.
Currently recovering from panic and retrying the task.

**Comment by jhendrixMSFT** [comment]
I think I understand what's going on here.  Does your code call `Recover()` from another goroutine?

**Comment by jhendrixMSFT** [comment]
So `Recover()` wasn't performing its work atomically, leaving a window between `Close()` and `newSessionAndLink()` where the lock wasn't held; I've fixed this in my commit.

I believe there's still one more race to fix though.  In `trySend()` we access `s.sender` without holding the read lock.  I'm a little hesitant to hold it while `s.sender.Send(ctx, msg)` is in flight though.

**Comment by johejo** [comment]
Yeah as you're concerned, I'm calling Send from multiple goroutines.

File: sender.go [hunk 3]
```diff
@@ -108,6 +112,11 @@ func (s *Sender) Close(ctx context.Context) error {
 	s.clientMu.Lock()
 	defer s.clientMu.Unlock()
 
+	return s.close(ctx)
+}
+
+// closes the connection.  callers *must* hold the client write lock before calling!
+func (s *Sender) close(ctx context.Context) error {
```
Comment (catalinaperalta) [inline sender.go:119]: Why is it necessary to move the functionality in Close() to close()?

Comment (jhendrixMSFT) [reply sender.go:119]: So that in `Recover()` we can `close()` and `newSessionAndLink()` while holding the write lock.

Comment (jhendrixMSFT) [reply sender.go:119]: In the previous implementation there was a window between `Close()` and `newSessionAndLink()` where the write lock wasn't held.  So you could `Close()` the connection then `Send()` would acquire the read lock however the connection hadn't been rebuilt yet.

**Comment by johejo** [comment]
Great!, Thank you! :+1:

**Comment by jhendrixMSFT** [comment]
Thanks for your help, sorry it took so long.

**Pull request merged**